File: src/novel_generator/evaluators/style.py
"""Style evaluator for Kakao Page style compliance."""

import re

from novel_generator.schema.novel import StyleGuide


class StyleEvaluator:
    """Evaluates chapter content for Kakao Page style compliance."""

    def __init__(self, style_guide: StyleGuide):
        self.style_guide = style_guide

    def evaluate(self, content: str) -> dict:
        """Evaluate content and return scores."""
        return {
            "dialogue_ratio": self._check_dialogue_ratio(content),
            "paragraph_length": self._check_paragraph_length(content),
            "sentence_length": self._check_sentence_length(content),
            "hook_ending": self._check_hook_ending(content),
            "overall_score": self._calculate_overall(content),
        }

    def _check_dialogue_ratio(self, content: str) -> dict:
        """Check dialogue to narration ratio."""
        # Count dialogue (text in quotes)
        dialogue_pattern = r'["\u201C\u201D]([^"\u201C\u201D]+)["\u201C\u201D]'
        dialogues = re.findall(dialogue_pattern, content)
        dialogue_chars = sum(len(d) for d in dialogues)

        total_chars = len(content)
        ratio = dialogue_chars / total_chars if total_chars > 0 else 0

        target = self.style_guide.dialogue_ratio
        score = 1.0 - min(abs(ratio - target) / target, 1.0)

        return {
            "actual_ratio": ratio,
            "target_ratio": target,
            "score": score,
            "pass": score >= 0.7,
        }

    def _check_paragraph_length(self, content: str) -> dict:
        """Check paragraph lengths."""
        paragraphs = [p.strip() for p in content.split("\n\n") if p.strip()]
        max_allowed = self.style_guide.max_paragraph_length

        violations = 0
        for para in paragraphs:
            # Count sentences (rough)
            sentences = re.split(r'[.!?]\s+', para)
            if len(sentences) > max_allowed:
                violations += 1

        score = 1.0 - (violations / len(paragraphs)) if paragraphs else 1.0

        return {
            "total_paragraphs": len(paragraphs),
            "violations": violations,
            "max_allowed_sentences": max_allowed,
            "score": score,
            "pass": score >= 0.8,
        }

    def _check_sentence_length(self, content: str) -> dict:
        """Check for short, punchy sentences."""
        sentences = re.split(r'[.!?]\s+', content)
        sentences = [s.strip() for s in sentences if s.strip()]

        # Target: most sentences under 50 chars for Korean
        short_threshold = 50
        short_count = sum(1 for s in sentences if len(s) <= short_threshold)

        ratio = short_count / len(sentences) if sentences else 1.0
        score = ratio  # Higher ratio of short sentences = better

        return {
            "total_sentences": len(sentences),
            "short_sentences": short_count,
            "short_ratio": ratio,
            "score": score,
            "pass": score >= 0.6,
        }

    def _check_hook_ending(self, content: str) -> dict:
        """Check if chapter ends with a hook."""
        # Get last paragraph
        paragraphs = [p.strip() for p in content.split("\n\n") if p.strip()]
        if not paragraphs:
            return {"score": 0.0, "pass": False, "reason": "No content"}

        last_para = paragraphs[-1]

        # Hook indicators
        hook_patterns = [
            r'[.]{3}$',  # Ends with ...
            r'[?!]$',  # Ends with ? or !
            r'그때',  # "At that moment"
            r'순간',  # "The moment"
            r'하지만',  # "But"
            r'그러나',  # "However"
            r'바로',  # "Right then"
        ]

        has_hook = any(re.search(p, last_para) for p in hook_patterns)

        return {
            "has_hook": has_hook,
            "last_paragraph": last_para[:100] + "..." if len(last_para) > 100 else last_para,
            "score": 1.0 if has_hook else 0.3,
            "pass": has_hook,
        }

    def _calculate_overall(self, content: str) -> float:
        """Calculate overall style score."""
        dialogue = self._check_dialogue_ratio(content)["score"]
        paragraph = self._check_paragraph_length(content)["score"]
        sentence = self._check_sentence_length(content)["score"]
        hook = self._check_hook_ending(content)["score"]

        # Weighted average
        weights = {
            "dialogue": 0.3,
            "paragraph": 0.2,
            "sentence": 0.2,
            "hook": 0.3,
        }

        overall = (
            dialogue * weights["dialogue"]
            + paragraph * weights["paragraph"]
            + sentence * weights["sentence"]
            + hook * weights["hook"]
        )

        return overall

File: web/src/lib/evaluators/style.ts
import type { StyleGuide } from "../schema/novel";

export interface StyleResult {
  dialogue_ratio: {
    actual_ratio: number;
    target_ratio: number;
    score: number;
    pass: boolean;
  };
  paragraph_length: {
    total_paragraphs: number;
    violations: number;
    max_allowed_sentences: number;
    score: number;
    pass: boolean;
  };
  sentence_length: {
    total_sentences: number;
    short_sentences: number;
    short_ratio: number;
    score: number;
    pass: boolean;
  };
  hook_ending: {
    has_hook: boolean;
    last_paragraph: string;
    score: number;
    pass: boolean;
  };
  overall_score: number;
}

export function evaluateStyle(
  content: string,
  styleGuide: StyleGuide,
): StyleResult {
  const dialogueResult = checkDialogueRatio(content, styleGuide.dialogue_ratio);
  const paragraphResult = checkParagraphLength(
    content,
    styleGuide.max_paragraph_length,
  );
  const sentenceResult = checkSentenceLength(content);
  const hookResult = checkHookEnding(content);

  const overall =
    dialogueResult.score * 0.3 +
    paragraphResult.score * 0.2 +
    sentenceResult.score * 0.2 +
    hookResult.score * 0.3;

  return {
    dialogue_ratio: dialogueResult,
    paragraph_length: paragraphResult,
    sentence_length: sentenceResult,
    hook_ending: hookResult,
    overall_score: overall,
  };
}

function checkDialogueRatio(content: string, target: number) {
  const dialoguePattern = /["\u201C\u201D]([^"\u201C\u201D]+)["\u201C\u201D]/g;
  let dialogueChars = 0;
  let match;
  while ((match = dialoguePattern.exec(content)) !== null) {
    dialogueChars += match[1].length;
  }
  const totalChars = content.length;
  const ratio = totalChars > 0 ? dialogueChars / totalChars : 0;
  const score = 1.0 - Math.min(Math.abs(ratio - target) / target, 1.0);
  return {
    actual_ratio: ratio,
    target_ratio: target,
    score,
    pass: score >= 0.7,
  };
}

function checkParagraphLength(content: string, maxAllowed: number) {
  const paragraphs = content.split("\n\n").filter((p) => p.trim());
  let violations = 0;
  for (const para of paragraphs) {
    const sentences = para.split(/[.!?]\s+/);
    if (sentences.length > maxAllowed) violations++;
  }
  const score =
    paragraphs.length > 0 ? 1.0 - violations / paragraphs.length : 1.0;
  return {
    total_paragraphs: paragraphs.length,
    violations,
    max_allowed_sentences: maxAllowed,
    score,
    pass: score >= 0.8,
  };
}

function checkSentenceLength(content: string) {
  const sentences = content.split(/[.!?]\s+/).filter((s) => s.trim());
  const shortThreshold = 50;
  const shortCount = sentences.filter(
    (s) => s.length <= shortThreshold,
  ).length;
  const ratio = sentences.length > 0 ? shortCount / sentences.length : 1.0;
  return {
    total_sentences: sentences.length,
    short_sentences: shortCount,
    short_ratio: ratio,
    score: ratio,
    pass: ratio >= 0.6,
  };
}

function checkHookEnding(content: string) {
  const paragraphs = content.split("\n\n").filter((p) => p.trim());
  if (paragraphs.length === 0)
    return { has_hook: false, last_paragraph: "", score: 0, pass: false };

  const lastPara = paragraphs[paragraphs.length - 1];
  const hookPatterns = [
    /\.{3}$/,           // 말줄임표
    /[?!]$/,            // 의문/감탄
    /["\u201C\u201D]/, // 대사로 끝남 (긴장감 있는 대사 후킹)
    /그때/,
    /순간/,
    /하지만/,
    /그러나/,
    /바로/,
    /아니/,
    /없었다\.$/,        // 부정적 종결 (불안감)
    /못했다\.$/,
    /몰랐다\.$/,
    /않았다\.$/,
    /있었다\.$/,        // 존재/발견 종결
    /보였다\.$/,
    /들렸다\.$/,
    /열렸다\.$/,
    /닫혔다\.$/,
  ];
  const hasHook = hookPatterns.some((p) => p.test(lastPara));

  return {
    has_hook: hasHook,
    last_paragraph:
      lastPara.length > 100 ? lastPara.slice(0, 100) + "..." : lastPara,
    score: hasHook ? 1.0 : 0.3,
    pass: hasHook,
  };
}
